### PR TITLE
Add optimistic locking support to etcd backend

### PR DESCRIPTION
**Pull request by rosstimothy**
Contributes to #30416.

### CODE REVIEW

File: lib/backend/etcdbk/etcd.go [hunk 1]
```diff
@@ -265,6 +265,7 @@ func New(ctx context.Context, params backend.Params, opts ...Option) (*EtcdBacke
 		CleanupInterval: utils.SeventhJitter(time.Minute * 2),
 	})
 	if err != nil {
+		cancel()
```
Comment (espadolini) [inline lib/backend/etcdbk/etcd.go:268]: If we're canceling the context here, shouldn't we also cancel it at other points in the initialization process?

File: lib/backend/etcdbk/etcd.go [hunk 2]
```diff
@@ -615,16 +617,33 @@ func (b *EtcdBackend) GetRange(ctx context.Context, startKey, endKey []byte, lim
...
 }
 
+func toBackendRevision(rev int64) string {
+	return strconv.FormatInt(rev, 10)
+}
+
+func fromBackendRevision(rev string) (int64, error) {
+	if rev == "" {
+		return 0, nil
+	}
```
Comment (espadolini) [inline lib/backend/etcdbk/etcd.go:637]: Why do we need an escape hatch for a blank revision string?

File: lib/backend/etcdbk/etcd.go [hunk 3]
```diff
@@ -771,11 +833,16 @@ func (b *EtcdBackend) Get(ctx context.Context, key []byte) (*backend.Item, error
...
+	value, err := unmarshal(kv.Value)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return &backend.Item{Key: key, Value: bytes, ID: kv.ModRevision, LeaseID: kv.Lease}, nil
+	return &backend.Item{
+		Key:      key,
+		Value:    value,
+		ID:       kv.ModRevision,
+		Revision: toBackendRevision(re.Header.Revision),
```
Comment (espadolini) [inline lib/backend/etcdbk/etcd.go:844]: If I'm reading the docs right, the `Revision` in the `ResponseHeader` for a range request is the revision number at the point where the request was processed, not the revision number for when the item was last modified (we expect different `ModRevision`s for different items in a `GetRange`, after all).

```suggestion
		Revision: toBackendRevision(kv.ModRevision),
```

File: lib/backend/etcdbk/etcd.go [hunk 4]
```diff
@@ -672,6 +695,42 @@ func (b *EtcdBackend) Update(ctx context.Context, item backend.Item) (*backend.L
...
+		if err := b.setupLease(ctx, item, &lease, &opts); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	start := b.clock.Now()
+	key := b.prependPrefix(item.Key)
+	rev, err := fromBackendRevision(item.Revision)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Comment (espadolini) [inline lib/backend/etcdbk/etcd.go:717]: We should do this before we set up the lease.

**Review by espadolini — APPROVED** [review]

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**